Move block cache from memory to files

LarryRuane opened the pull request:
Fixes #182. Improve performance to the wallets of fetching blocks, all 300k mainnet blocks since Sapling returned in 20 seconds, using grpcurl and `GetBlockRange` locally (almost as fast as setting `-cache-size` to 300k in the current version), but the `lightwalletd` restart time is instantaneous, instead of about 40 minutes (to load the memory cache back to Sapling). Also, since there is no memory caching, the memory footprint is much smaller (28 MB) than the current version when configured to maintain a full cache (about 223 MB). The disk space requirement of this version increases from zero to about 50mb (block data can be stored more efficiently on disk than in a memory cache). The `-cache-size` command-line argument has been removed since the cache must always go back to the first Sapling block.

LarryRuane commented:
@adityapk00, please take a look at this, if you have time (I couldn't add you as a reviewer but would love your review.)

pacu commented:
@LarryRuane I checked out the branch, I didn't notice it syncing from sapling activation. it synced from  block 715000 on

```
{"app":"frontend-grpc","level":"info","msg":"Got sapling height 280000 chain test branchID 2bb40e60","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg":"Starting gRPC server on 127.0.0.1:9067","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg":"Ingestor adding block to cache: 715651","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg":"Ingestor adding block to cache: 715652","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg":"Ingestor adding block to cache: 715653","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg":"Ingestor adding block to cache: 715654","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg":"Ingestor adding block to cache: 715655","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg":"Ingestor adding block to cache: 715656","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg":"Ingestor adding block to cache: 715657","time":"2020-02-10T13:54:14-03:00"}
{"app":"frontend-grpc","level":"info","msg
```

this is how I launched it 
```
./server --no-tls-very-insecure=true --conf-file ~/Library/Application\ Support/Zcash/zcash.conf --log-file server.log --bind-addr 127.0.0.1:9067
```

My Zcash.conf

````
addnode=testnet.z.cash
testnet=1
server=1
rpcuser=[REDACTED]
rpcpassword=[REDACTED]
rpcbind=127.0.0.1
rpcport=8232
experimentalfeatures=1
txindex=1
insightexplorer=1
              
````

LarryRuane commented:
I changed this to WIP because it needs to handle file corruption. The good news is that both files are simply a sequence of per-block data ordered by height. If file corruption is detected, we can truncate the two `db-*` files at the offset corresponding to that block height, and continue fetching blocks starting from that height. In the worst case (corruption at the start of either file), we will just download all blocks (as if the files didn't exist). So recovery is easy.

The detection of corruption is also easy. To detect corruption:
- As we're reading the `db-*-lengths` file during startup, sanity-check the block lengths. There is a relatively narrow range of possible values.
- Precede each block by its checksum or hash. It could be the actual block hash, but that would mean the `db-*-blocks` file would be about 32 MB larger (with a million blocks), a pretty large fraction of the current 50 MB. Since this doesn't have to be a cryptographically-secure hash, we can get by with a 4 or 8-byte hash, as documented [here](https://golang.org/pkg/hash/#pkg-subdirectories). With even a 4-byte checksum, the per-block probability of undetected corruption is one in 4 billion.

Startup time would be delayed too much if we verify the checksums at that time, so I propose to verify the checksum each time we read a block and its checksum from disk (still a single read). If corruption is detected, then take the action described above (truncate both files at that point, fetch the requested and subsequent blocks from `zcashd`). We should verify the checksum on each block read anyway, in case the file is corrupted after startup.

It will be straightforward to unit-test these changes.

defuse commented on common/cache.go:
```diff
@@ -10,87 +13,169 @@ import (
+	binary.LittleEndian.PutUint32(b, uint32(len(data)))
+	_, err = c.lengthsFile.Write(b)
+	if err != nil {
+		Log.Fatal("lengths write failed: ", err)
```
Bug: At this point a block has been written to the block file, but a length has not been added to the length file. `NewBlockCache` will not detect the problem, because it uses the number of entries in the length file to determine how many blocks to read from the blocks file. `NewBlockCache` doesn't check that there aren't *extra* blocks in the file.

The following could happen:

1. We try to add a block A, at height N, and reach this path due to some transient error. The blocks file has N blocks ending in A, but the length file only has N-1 lengths.
2. Lightwalletd gets restarted, `NewBlockCache` doesn't notice the error, lightwalletd serves blocks up to N-1.
3. We try to add a block B, at height N (say there was a reorg), but the transient error is resolved and this code works. Now, before lightwalletd restarts, it'll serve block B at height N, but after it restarts, it will serve block A at height N.

(Assuming blocks A and B are the same length, otherwise it might crash upon trying to read the block)

One way to fix this would be to add a "transaction in progress" flag to each file, and modify them by:

1. Set the "transaction in progress" flag in both files. Sync to guarantee that has been written to disk.
2. Write the block contents and then the length. Sync to guarantee it has all been written to disk.
3. Clear the "transaction in progress" flag in both files.
4. Upon loading the file, make sure both "transaction in progress" flags are cleared, or assume the database is corrupted.

(I guess it could be a flag in its own file or in just one of the files).

LarryRuane replied:
> `NewBlockCache` doesn't check that there aren't _extra_ blocks in the file
What if `NewBlockCache` unconditionally just truncates the blocks file to the length it should have based on the lengths file? Would that fix this? 

> but after it restarts, it will serve block A at height N.
Is that because block B didn't get synced to disk? One thing I was trying to avoid is syncing files in `Add()` because while we're syncing with the blockchain, we're calling `Add()` for each block, so performance matters. (But notice we do `Sync()` before sleeping in `BlockIngestor()`.) I guess I'm thinking that with the checksum improvement, any corruption that occurs (that could have been prevented by syncing) will be detected and the bad data overwritten (by fetching from `zcashd` again).

Hmm, here's a thought: When we restart, instead of fetching from exactly where (we think) we left off, back up 100 blocks or something (go through exactly the reorg code), and resume from there. The time to re-download 100 blocks is very small. This would overwrite any corruption near the end of the files, which is by far the most likely place for it to be (because that's where writing occurs).

defuse commented on common/cache.go:
```diff
@@ -10,87 +13,169 @@ import (
-		delete(c.m, c.firstBlock)
-		c.firstBlock++
+	if err := c.lengthsFile.Truncate(int64(4 * newCacheLen)); err != nil {
+		Log.Fatal("truncate failed: ", err)
```
This could also leave the database files in a state where there's more blocks in the blocks file than there are lengths in the lengths file and lead to similar bugs as above. The same fix works here too.

LarryRuane commented on common/common.go:
```diff
@@ -133,26 +137,24 @@ func BlockIngestor(cache *BlockCache, startHeight int, rep int) {
-			reorgCount++
-			if reorgCount > 10 {
+			height = c.Reorg(height - 2)
+			reorgCount += 2
```
It's 2 because I changed the definition of `reorgCount` to blocks instead of attempts. That makes the code immediately below make more sense:
```
if reorgCount > 100 {
	Log.Fatal("Reorg exceeded max of 100 blocks! Help!")
}
```

LarryRuane commented on common/cache.go:
```diff
@@ -10,87 +13,169 @@ import (
-		MaxEntries: maxEntries,
-		m:          make(map[int]*blockCacheEntry),
+// Note if you call this with a different height than was used
+// to create the db files, first delete those files (fixme)
```
When starting up, we could fetch just this very first block from `zcashd` and make sure it matches what's in the file. (If it doesn't, then we'd truncate both files to zero and re-download all blocks, as we should; if it does match, we go on with our startup without checking any more blocks.)

LarryRuane commented on common/cache.go:
```diff
@@ -10,87 +13,169 @@ import (
-		delete(c.m, i)
+	_, err = c.blocksFile.Write(data)
+	if err != nil {
+		Log.Fatal("blocks write failed: ", err)
```
Yes, checksum should fix this.

LarryRuane commented:
@defuse, I think I addressed all your concerns. If any data corruption is detected, the db files are automatically truncated to zero-length, and the ingestor begins downloading from the beginning. During that time, `GetBlock()` requests are satisfied but will be slower (just like during any initial block download). The server tries very hard to recover gracefully instead of crashing.

If you would like to test out the corruption handling, I added a tiny program, `testtools/zap.go`, which can "zap" (increment) a specified byte in a specified file. I ran that on a byte near the start of the `db-main-blocks` file while the server was running, and then requested that block (using `grpcurl ... GetBlock`), and it got the correct response (because it reverted to asking `zcashd`), but the lightwalletd truncated both of its files to empty, and began (or restarted) downloading from the beginning (you can see that this happened in `server.log`). Or you can zap various offsets in the `blocks` or `lengths` files while lightwalletd isn't running, and you will see that when you start it, it detects the corruption and starts over. (The `lengths` file isn't read after startup.) Except, note that it doesn't verify every block during startup (that would take too long); it does verify the first block, and it verifies blocks as they are read (`GetBlock` and `GetBlockRange` gRPCs).

I did a lot of manual testing, but haven't written any corruption-detection and recovery unit tests yet; I really don't have time in this sprint and would like to get these changes in, so I will get to it later (if you don't mind).

defuse commented on common/cache.go:
```diff
@@ -13,90 +18,300 @@ import (
+	return c.latestHash != nil && !bytes.Equal(c.latestHash, prevhash)
+}
+
+// Start over, so that all blocks get re-downloaded (file corruption)
```
If we ever determine that there's file corruption, it means there's a bug somewhere in the code or some really unlikely I/O error occurred, so we might want to leave a copy of the corrupted files on disk so we can investigate. That doesn't block this PR but we should open a ticket to do it at some point.

defuse replied:
Here's the ticket: https://github.com/zcash/lightwalletd/issues/233

defuse commented on common/cache.go:
```diff
@@ -13,90 +18,300 @@ import (
+}
+
+// Start over, so that all blocks get re-downloaded (file corruption)
+func (c *BlockCache) truncZero() {
```
Requested change: This should always generate a log message, it looks like there's a log message in every case of detecting corruption, but it would be nice to have a single consistent log message that sysadmins can set up their alerting to detect. It would also be great to document the log message in the README, so everyone running this in a production environment is aware this can happen and what to look for in the logs.

defuse commented on common/cache.go:
```diff
@@ -13,90 +18,300 @@ import (
+			// start over
+			Log.Warningln("lengths file has impossible value, re-downloading")
+			c.truncZero()
+			offset = 0
```
Nice defense-in-depth check!

defuse commented on common/cache.go:
```diff
@@ -13,90 +18,300 @@ import (
+		offset += int64(length) + 8
+		c.starts = append(c.starts, offset)
+		if i == 0 {
+			// first block, sanity check it has Sapling activation height
```
Awesome!

defuse commented on common/cache.go:
```diff
@@ -13,90 +18,300 @@ import (
+		c.nextBlock++
+	}
+
+	// Make sure the files have the correct length (if unclean shutdown)
```
Requested change: If the truncation actually would change the file, there should be some kind of error log generated here.

defuse replied:
An easy way to do that might be to try reading a byte from the blocks file and check if it's at EOF. The length file will always have the correct length unless we take `break` out of the loop, in which case the file will be truncated to zero length anyways.
